src/main/groovy/lang/NumberRange.java

```diff
@@ -0,0 +1,536 @@
+        }
+        reverse = areReversed(from, to);
+        Number tempFrom;
+        Number tempTo;
+        if (reverse) {
+            tempFrom = to;
+            tempTo = from;
+        } else {
+            tempFrom = from;
+            tempTo = to;
+        }
+        if (tempFrom instanceof Short) {
+            tempFrom = tempFrom.intValue();
+        } else if (tempFrom instanceof Float) {
+            tempFrom = tempFrom.doubleValue();
+        }
+        if (tempTo instanceof Short) {
+            tempTo = tempTo.intValue();
+        } else if (tempTo instanceof Float) {
+            tempTo = tempTo.doubleValue();
+        }
+
+        if (tempFrom instanceof Integer && tempTo instanceof Long) {
+            tempFrom = tempFrom.longValue();
+        } else if (tempTo instanceof Integer && tempFrom instanceof Long) {
+            tempTo = tempTo.longValue();
+        }
+
+        this.from = (Comparable) tempFrom;
+        this.to = (Comparable) tempTo;
+        this.stepSize = stepSize == null ? 1 : stepSize;
+        this.inclusive = inclusive;
+    }
+
+    /**
+     * For a NumberRange with step size 1, creates a new NumberRange with the same
+     * <code>from</code> and <code>to</code> as this NumberRange
+     * but with a step size of <code>stepSize</code>.
+     *
+     * @param stepSize the desired step size
+     * @return a new NumberRange
+     */
+    public <T extends Number & Comparable> NumberRange by(T stepSize) {
+        if (!Integer.valueOf(1).equals(this.stepSize)) {
+            throw new IllegalStateException("by only allowed on ranges with original stepSize = 1 but found " + this.stepSize);
+        }
+        return new NumberRange(comparableNumber(from), comparableNumber(to), stepSize, inclusive);
+    }
+
+    @SuppressWarnings("unchecked")
+    /* package private */ static <T extends Number & Comparable> T comparableNumber(Comparable c) {
+        return (T) c;
+    }
+
+    @SuppressWarnings("unchecked")
+    /* package private */ static <T extends Number & Comparable> T comparableNumber(Number n) {
+        return (T) n;
+    }
+
+    private static boolean areReversed(Number from, Number to) {
+        try {
+            return compareGreaterThan(from, to);
+        } catch (ClassCastException cce) {
+            throw new IllegalArgumentException("Unable to create range due to incompatible types: " + from.getClass().getSimpleName() + ".." + to.getClass().getSimpleName() + " (possible missing brackets around range?)", cce);
+        }
+    }
+
+    public boolean equals(Object that) {
+        return (that instanceof NumberRange) ? equals((NumberRange) that) : super.equals(that);
+    }
+
+    /**
+     * Compares an {@link NumberRange} to another {@link NumberRange}.
+     *
+     * @param that the object to check equality with
+     * @return <code>true</code> if the ranges are equal
+     */
+    public boolean equals(NumberRange that) {
```

Comment:
Maybe too simplistic but wonder if following would work:

```
    @Override
    public int hashCode() {
        int result = 17;
        result = result * 31 + (reverse ? 1 : 0);
        result = result * 31 + from.hashCode();
        result = result * 31 + to.hashCode();
        result = result * 31 + stepSize.hashCode();
        return result;
    }
```
That would mean the following would have different hash codes, but since inclusive isn't used in the equals they aren't equal.
```
Range n1 = new NumberRange(1L, 5L)
Range n2 = new NumberRange(1L, 6L, false)
```